internal/metrics/remote_write.go

```diff
@@ -0,0 +1,394 @@
+	config               RemoteWriteConfiguration
+	tsdb                 API
+	LastWrite            time.Time
+	lastWriteFile        string
+	waitInterval         time.Duration
+	RangeDuration        time.Duration
+	RequestNumSamples    int
+	RequestRetryInterval time.Duration
+	Client               WriteClient
+}
+
+type RemoteWriteConfiguration struct {
+	Url     string // empty URL means that feature is disabled
+	Timeout time.Duration
+}
+
+func NewRemoteWrite(dataDir, deviceID string, tsdbInstance API) *RemoteWrite {
+	waitInterval, _ := time.ParseDuration(os.Getenv("REMOTEWRITE_WAIT_INTERVAL"))
+	if waitInterval == 0 {
+		waitInterval = defaultWaitInterval
+	}
+	RangeDuration, _ := time.ParseDuration(os.Getenv("REMOTEWRITE_RANGE_DURATION"))
+	if RangeDuration == 0 {
+		RangeDuration = defaultRequestDuration
+	}
+	requestNumSamples, _ := strconv.Atoi(os.Getenv("REMOTEWRITE_REQUEST_NUM_SAMPLES"))
+	if requestNumSamples == 0 {
+		requestNumSamples = defaultRequestNumSamples
+	}
+	requestRetryInterval, _ := time.ParseDuration(os.Getenv("REMOTEWRITE_REQUEST_RETRY_INTERVAL"))
+	if requestRetryInterval == 0 {
+		requestRetryInterval = defaultRequestRetryInterval
+	}
+
+	newRemoteWrite := RemoteWrite{
+		deviceID:             deviceID,
+		tsdb:                 tsdbInstance,
+		waitInterval:         waitInterval,
+		RangeDuration:        RangeDuration,
+		RequestNumSamples:    requestNumSamples,
+		RequestRetryInterval: requestRetryInterval,
+		lastWriteFile:        path.Join(dataDir, LastWriteFileName),
+	}
+
+	lastWriteBytes, err := ioutil.ReadFile(newRemoteWrite.lastWriteFile)
+	if err == nil {
+		if len(lastWriteBytes) != 0 {
+			i, err := strconv.ParseInt(string(lastWriteBytes), 10, 64)
+			if err == nil {
+				newRemoteWrite.LastWrite = time.Unix(i/1000000000, i%1000000000)
+			} else {
+				log.Error(err)
+			}
+		}
+	} else if !os.IsNotExist(err) {
+		log.Error(err)
+	}
+
+	return &newRemoteWrite
+}
+
+// RemoteRecoverableError
+// 'remote.RecoverableError' fields are private and can't be instantiated during testing. Therefore, we use this wrapper
+type RemoteRecoverableError struct {
+	error
+}
+
+func (e RemoteRecoverableError) Error() string {
+	if e.error != nil {
+		return e.Error()
+	} else {
+		return ""
+	}
+}
+
+// WriteClient
+// an interface used for testing
+//go:generate mockgen -package=metrics -destination=write_client_mock.go . WriteClient
+type WriteClient interface {
+	Write(context.Context, []byte) error
+}
+
+type writeClient struct {
+	client remote.WriteClient
+}
+
+func (w *writeClient) Write(ctx context.Context, data []byte) error {
+	err := w.client.Store(ctx, data)
+	switch err.(type) {
+	case remote.RecoverableError:
+		return RemoteRecoverableError{err}
+	default:
+		return err
+	}
+}
+
+func (r *RemoteWrite) Update(config models.DeviceConfigurationMessage) error {
+	// the following two variables are only here until we implement reading form device configuration
+	serverURL := os.Getenv("REMOTEWRITE_SERVER_URL")
```

Comment:
this shouldn't part of the config for the device? getting from the API?

Reply:
in next story it will come from device config. see the comment in code

Reply:
once PR https://github.com/project-flotta/flotta-operator/pull/148 is merged I will fix